### ISSUE TITLE
[BUG] Top-level code isn't enforcing ownership and mutability

### ISSUE THREAD

**Issue by nmsmith** [modular/modular#15]
## Bug Description

The program below crashes with a double-free error.

### Steps to Reproduce

Run the following program in a notebook:
```
from String import String

struct Foo:
    var x: String
    fn __init__(self&, owned x: String):
        self.x = x
    fn __moveinit__(self&, owned existing: Self):
        self.x = existing.x

fn bar(owned foo: Foo) -> Int:
    return 2

let foo = Foo("test")
print(bar(foo^))
print(bar(foo^))
```

### Context

<details>
  <summary>Click to expand</summary>
11:freezer:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
10:pids:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
9:hugetlb:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
8:net_cls,net_prio:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
7:blkio:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
6:cpu,cpuacct:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
5:cpuset:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
4:devices:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
3:perf_event:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
2:memory:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
1:name=systemd:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod58ead308_6b5e_4f48_a483_33eff7b8c09b.slice/cri-containerd-bf25cf0ee968cddde56ce666988b8ec79651ef4adadc6ffc15ae99baf64d98d8.scope
0::/
</details>

**Comment by Mogball** [comment]
Thanks for the bug report! This seems specifically like a notebook environment bug, as it is occurring in top-level code (trying to do the same inside a function definition will cause a compilation error). @River707 I guess we need to track liveliness?

**Comment by lattner** [comment]
Yep, this is workbook toplevel code specific, thanks for reporting it!

**Comment by gabrieldemarmiesse** [comment]
Doing some triaging here, with Mojo 2024.1.0. I updated the example to reproduce the bug to the latest syntax:

```
struct Foo:
    var x: String
    fn __init__(inout self, owned x: String):
        self.x = x
    fn __moveinit__(inout self, owned existing: Self):
        self.x = existing.x

fn bar(owned foo: Foo) -> Int:
    return 2

var foo = Foo(String("test"))
print(bar(foo^))
print(bar(foo^))
```
We now get this output:

```
error: Expression [1]:12:14: cannot consume indirect references to values
print(bar(foo^))
             ^

error: Expression [1]:13:14: cannot consume indirect references to values
print(bar(foo^))
             ^

2
2
```
So we're getting errors in the output, but the notebook cell is still considered successful and the code is still executed. That's a different bug, but still a bug.

**Comment by goldiegadde** [comment]
testing
